scipy/signal/filter_design.py

```diff
@@ -328,22 +345,77 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
     >>> plt.show()
 
     """
-    b, a = map(atleast_1d, (b, a))
-    if whole:
-        lastpoint = 2 * pi
-    else:
-        lastpoint = pi
+    b = atleast_1d(b)
+    a = atleast_1d(a)
+    ndim = max(b.ndim, a.ndim)
+    if not -ndim <= axis < ndim:
+        raise ValueError('axis=%s invalid for ndim=%s' % (axis, ndim))
+    if axis >= 0:
+        axis = axis - ndim  # the way it will broadcast
+    try:
+        a_len = a.shape[axis]
+    except IndexError:
+        a_len = 1
+    try:
+        b_len = b.shape[axis]
+    except IndexError:
+        b_len = 1
+
     if worN is None:
-        N = 512
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
-    elif isinstance(worN, int):
-        N = worN
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
+        worN = 512
+
+    h = None
+    try:
+        worN = operator.index(worN)
+    except TypeError:  # not int-like
+        w = atleast_1d(worN)
     else:
-        w = worN
-    w = atleast_1d(w)
-    zm1 = exp(-1j * w)
-    h = polyval(b[::-1], zm1) / polyval(a[::-1], zm1)
+        if worN <= 0:
+            raise ValueError('worN must be positive, got %s' % (worN,))
+        lastpoint = 2 * pi if whole else pi
+        w = np.linspace(0, lastpoint, worN, endpoint=False)
+        min_size = max(a_len, b_len)
+        if worN >= min_size and fftpack.next_fast_len(worN) == worN:
+            # if worN is fast, 2 * worN will be fast, too,
+            # so no need to check that one
```

Comment:
Are there any numerical error issues with dividing by a spectrum? (I tried resonant 2nd-order filters so `b` is finite and `a` is close to zero, but it still worked fine.)